### ISSUE TITLE
Sessions add a lot of memory overhead

### ISSUE THREAD

**Issue by mperham** [bugsnag/bugsnag-ruby#536]
Hi, I'm trying to create a small Ruby service and find that bugsnag is 50% of the loaded gem overhead.  Here's what `derailed_benchmarks` says:

```
$ bundle exec derailed bundle:mem
TOP: 10.3594 MiB
  rubydns: 5.2969 MiB
    async/dns: 5.2852 MiB
      async/io: 1.6133 MiB
        openssl: 1.4297 MiB (Also required by: net/https)
          openssl/ssl: 0.7852 MiB
          openssl.so: 0.5859 MiB
      async/io/tcp_socket: 0.5469 MiB
  bugsnag: 5.0195 MiB
    bugsnag/session_tracker: 4.2734 MiB
      concurrent: 4.2734 MiB
        concurrent/executors: 0.9141 MiB
          concurrent/executor/timer_set: 0.4648 MiB
        concurrent/atomics: 0.7188 MiB (Also required by: concurrent/executor/simple_executor_service)
        concurrent/promises: 0.3867 MiB
        concurrent/configuration: 0.3242 MiB (Also required by: concurrent/scheduled_task, concurrent/options, and 2 others)
    bugsnag/delivery/synchronous: 0.418 MiB
      net/https: 0.4102 MiB
        net/http: 0.4063 MiB
```

Is it possible to lazy load the session code or its `concurrent` dependency to avoid this overhead (or load just the part of concurrent that you need)?  My usage of Bugsnag is literally `at_exit { Bugsnag.notify $! if $! }` so I don't believe I need sessions.

**Comment by snmaynard** [comment]
You could probably move the `require` into the `Bugsnag::SessionTracker#initialize` method and then if you aren't using the session tracker it shouldn't bump the memory usage. Haven't had a chance to test it but maybe it will solve your issue?

**Comment by mperham** [comment]
Looks similar to me in 6.12.1:

```
  bugsnag: 4.4844 MiB
    bugsnag/session_tracker: 3.0664 MiB
      concurrent: 3.0195 MiB
        concurrent/promises: 0.7773 MiB
        concurrent/executors: 0.4844 MiB
          concurrent/executor/timer_set: 0.3125 MiB
    bugsnag/delivery/synchronous: 0.625 MiB
      net/https: 0.5898 MiB
        /Users/mikeperham/.rubies/ruby-2.6.3/lib/ruby/2.6.0/net/http: 0.5898 MiB
    bugsnag/report: 0.3594 MiB
```

**Comment by mattdyoung** [comment]
@mperham 

Did you try @snmaynard's suggested tweak to the source code above to confirm if that makes a difference?
 https://github.com/bugsnag/bugsnag-ruby/issues/536#issuecomment-486867060

**Comment by mperham** [comment]
@mattdyoung I don't think I was clear.  I have this Gemfile:

```ruby
source 'https://rubygems.org'

gem 'rubydns'
gem 'bugsnag'
gem 'derailed_benchmarks'
```

I `bundle` and then run `bundle exec derailed bundle:mem`. I'm not executing any of my own Ruby code.  The session code is required just by calling `gem 'bugsnag'`.

**Comment by mattdyoung** [comment]
My apologies, I should have explained more clearly. @snmaynard was suggesting you try editing the code from the bugsnag-ruby repository and create your own fork which doesn't load the `session_tracker` until it's used.

You could try this by moving this require:
https://github.com/bugsnag/bugsnag-ruby/blob/0466dd2a654cc1adbd1e1bb2fb7ba128fae3ba85/lib/bugsnag.rb#L10
to the session_tracker `initialize` method:
https://github.com/bugsnag/bugsnag-ruby/blob/2fbc548ffd4e115409f48ba472dd54b3cc5384da/lib/bugsnag/session_tracker.rb#L29

Running with that fork might work for your needs.

If we have interest from other customers for this sort of configuration we may consider building it into the official library in the future.

Thanks!

**Comment by phillipsam** [comment]
Hi @mperham - an update for this has been released in [v6.15.0](https://github.com/bugsnag/bugsnag-ruby/releases/tag/v6.15.0)
